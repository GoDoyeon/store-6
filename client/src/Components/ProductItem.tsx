import React from 'react';
import styled from 'styled-components';
import { useState } from 'react';

interface Props {
    idx: number;
    thumbnail: string;
    title: string;
    price: string;
}

const ProductItem: React.FC<Props> = ({
    idx,
    thumbnail,
    title,
    price
}) => {

  const [isHoverImage, setIsHoverImage] = useState<boolean>(false);

  const handleOnMouseEnter = ():void => { 
    setIsHoverImage(true);
  }

  const handleOnMouseLeave = ():void => { 
    setIsHoverImage(false);
  }

  return (
    <ProductItemContainer onMouseEnter={handleOnMouseEnter} onMouseLeave={handleOnMouseLeave}>
      <ProductImage src={thumbnail}></ProductImage>
        {isHoverImage && 
        <ProductInfoConatiner>
          <ProductTitle>{title}</ProductTitle>
          <ProductPrice>{price}</ProductPrice>
      </ProductInfoConatiner>
      }
    </ProductItemContainer>
  );
};

const ProductItemContainer = styled.div`
  position: relative;
  height: 300px;
  width: 230px;
  transition: all 300ms;
  &: hover{
    transform: scale( 1.08 );
  }
`;

const ProductImage = styled.img`
  height: 300px;
  width: 230px;
  border-radius: 10px;
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25)) drop-shadow(2px 2px 12px rgba(0, 0, 0, 0.15));
  transition: all 300ms;
  ${ProductItemContainer}:hover & {
    filter: brightness(60%);
    transform: scale( 1.08 );
    box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25), 2px 2px 12px rgba(0, 0, 0, 0.15)
  }
`;

const ProductInfoConatiner = styled.div`
  position: absolute;
  left: rem;
  bottom: 0.7rem;
`;

const ProductTitle = styled.div`
  width: 204px;
  height: 32px;
  font-size: 24px;
  color: #FFFFFF;
`;

const ProductPrice = styled.div`
  width: 179px;
  height: 18px;
  font-size: 18px;
  color: #FFFFFF;
`;

export default ProductItem;
