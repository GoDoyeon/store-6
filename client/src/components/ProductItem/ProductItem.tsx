import React, { useState } from 'react';
import styled from 'styled-components';

interface Props {
  idx: number;
  thumbnail: string;
  title: string;
  price: string;
}

const ProductItemContainer = styled.div`
  position: relative;
  height: 300px;
  width: 230px;
  transition: all 300ms;
  &:hover {
    transform: scale(1.08);
  }
`;

const ProductImage = styled.img`
  height: 300px;
  width: 230px;
  border-radius: 10px;
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25))
    drop-shadow(2px 2px 12px rgba(0, 0, 0, 0.15));
  transition: all 300ms;
  ${ProductItemContainer}:hover & {
    filter: brightness(60%);
    transform: scale(1.08);
    box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25),
      2px 2px 12px rgba(0, 0, 0, 0.15);
  }
`;

const ProductInfoConatiner = styled.div`
  position: absolute;
  left: 0.5rem;
  bottom: 0.7rem;
`;

const ProductTitle = styled.div`
  width: 204px;
  height: 32px;
  font-size: 24px;
  color: #ffffff;
`;

const ProductPrice = styled.div`
  width: 179px;
  height: 18px;
  font-size: 18px;
  color: #ffffff;
`;

const ProductItem: React.FC<Props> = ({ idx, thumbnail, title, price }) => {
  const [isProductImageHovered, setIsProductImageHovered] =
    useState<boolean>(false);

  const handleOnMouseEnter = (): void => {
    setIsProductImageHovered(true);
  };

  const handleOnMouseLeave = (): void => {
    setIsProductImageHovered(false);
  };

  return (
    <ProductItemContainer
      onMouseEnter={handleOnMouseEnter}
      onMouseLeave={handleOnMouseLeave}
    >
      <ProductImage src={thumbnail} />
      {isProductImageHovered && (
        <ProductInfoConatiner>
          <ProductTitle>{title}</ProductTitle>
          <ProductPrice>{price}</ProductPrice>
        </ProductInfoConatiner>
      )}
    </ProductItemContainer>
  );
};

export default ProductItem;
