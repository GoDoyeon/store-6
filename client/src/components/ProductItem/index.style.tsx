import styled from 'styled-components';

export const ProductItemWrapper = styled.li`
  position: relative;
  height: 300px;
  width: 230px;
  cursor: pointer;
`;

export const ProductImage = styled.img<{ delayedIsHovered: boolean }>`
  height: 300px;
  width: 230px;
  border-radius: 10px;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25), 2px 2px 12px rgba(0, 0, 0, 0.15);
  transform: ${({ delayedIsHovered }) =>
    delayedIsHovered ? `scale(1.08)` : 'none'};
  transition: all 300ms ease-in-out;
`;

export const ProductInfoWrapper = styled.div<{ delayedIsHovered: boolean }>`
  position: absolute;
  left: 15px;
  bottom: 20px;
  transition: opacity 300ms ease-in-out;
  opacity: ${({ delayedIsHovered }) => (delayedIsHovered ? '1' : '0')};
  text-shadow: 4px 4px 10px rgba(0, 0, 0, 0.65);
  color: #fff;
`;

export const ProductTitle = styled.span`
  display: block;
  width: 169px;
  font-family: 'BM Hanna' !important;
  font-size: 20px;
  margin-bottom: 10px;
`;

export const ProductPrice = styled.span`
  width: 179px;
  font-family: 'BM Hanna' !important;
  font-size: 16px;
`;
