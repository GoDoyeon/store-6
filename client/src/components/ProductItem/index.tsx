import React, { useState } from 'react';
import {
  ProductItemWrapper,
  ProductImage,
  ProductInfoWrapper,
  ProductTitle,
  ProductPrice,
} from './index.style';

interface Props {
  thumbnail: string;
  title: string;
  price: number;
}

const ProductItem: React.FC<Props> = ({ thumbnail, title, price }) => {
  const [isProductImageHovered, setIsProductImageHovered] =
    useState<boolean>(false);

  const handleOnMouseEnter = (): void => setIsProductImageHovered(true);
  const handleOnMouseLeave = (): void => setIsProductImageHovered(false);

  return (
    <ProductItemWrapper
      onMouseEnter={handleOnMouseEnter}
      onMouseLeave={handleOnMouseLeave}
    >
      <ProductImage src={thumbnail} />
      {isProductImageHovered && (
        <ProductInfoWrapper>
          <ProductTitle>{title}</ProductTitle>
          <ProductPrice>{price}</ProductPrice>
        </ProductInfoWrapper>
      )}
    </ProductItemWrapper>
  );
};

export default ProductItem;
