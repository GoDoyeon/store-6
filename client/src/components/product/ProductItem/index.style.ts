import styled from 'styled-components';

export const ProductItemContainer = styled.div`
  position: relative;
  height: 300px;
  width: 230px;
  transition: all 300ms;
`;

export const ProductImage = styled.img`
  height: 300px;
  width: 230px;
  border-radius: 10px;
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25))
    drop-shadow(2px 2px 12px rgba(0, 0, 0, 0.15));
  transition: all 300ms;
  ${ProductItemContainer}:hover & {
    filter: brightness(60%);
    transform: scale(1.08);
    box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25),
      2px 2px 12px rgba(0, 0, 0, 0.15);
  }
`;

export const ProductInfoConatiner = styled.div`
  position: absolute;
  left: 0.5rem;
  bottom: 0.7rem;
`;

export const ProductTitle = styled.div`
  width: 204px;
  height: 32px;
  font-size: 24px;
  color: #ffffff;
`;

export const ProductPrice = styled.div`
  width: 179px;
  height: 18px;
  font-size: 18px;
  color: #ffffff;
`;
