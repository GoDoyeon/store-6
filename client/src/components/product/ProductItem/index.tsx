import React, { useCallback, useState } from 'react';
import { formatPrice } from '~/utils/formatPrice';
import ProductItemWrapper from '~/components/product/ProductItemWrapper';
import {
  ProductInfoWrapper,
  ProductTitle,
  ProductPrice,
  ProductLikeButtonWrapper,
} from './index.style';
import ProductImage from '../ProductImage';
import ProductLikeButton from '../ProductLikeButton';
import debounce from '~/utils/debounce';

interface Props {
  idx: number;
  thumbnail: string;
  title: string;
  price: number;
  onClick?: (idx: number) => void;
  onClickLike?: (idx: number) => void;
  isLike?: boolean;
  isLikeItem?: boolean;
}

const ProductItem: React.FC<Props> = ({
  idx,
  thumbnail,
  title,
  price,
  onClick,
  onClickLike,
  isLike = false,
  isLikeItem = false,
}) => {
  const [isHovered, setIsHovered] = useState<boolean>(false);

  const DELAYED_TIME = 100;
  const delayMouseEnter = useCallback(
    debounce(() => setIsHovered(true), DELAYED_TIME),
    [],
  );

  const delayMouseLeave = useCallback(
    debounce(() => setIsHovered(false), DELAYED_TIME),
    [],
  );

  const onClickHandler = useCallback(() => {
    if (onClick) onClick(idx);
  }, [onClick, idx]);

  const onClickLikeHandler = useCallback(
    (e: React.MouseEvent<HTMLButtonElement>) => {
      e.stopPropagation();

      if (onClickLike) onClickLike(idx);
    },
    [onClickLike, idx],
  );

  return (
    <ProductItemWrapper
      onMouseEnter={delayMouseEnter}
      onMouseLeave={delayMouseLeave}
      onClick={onClickHandler}
    >
      <ProductImage
        src={thumbnail}
        autoHover={false}
        referrerPolicy="no-referrer"
        isHovered={isLikeItem ? false : isHovered}
      />
      {isLikeItem && (
        <ProductLikeButtonWrapper>
          <ProductLikeButton
            isLike={isLike}
            onClick={onClickLikeHandler}
            fillLineWhenHover
          />
        </ProductLikeButtonWrapper>
      )}
      <ProductInfoWrapper isHovered={isHovered}>
        <ProductTitle>{title}</ProductTitle>
        <ProductPrice>{formatPrice(price)}</ProductPrice>
      </ProductInfoWrapper>
    </ProductItemWrapper>
  );
};

export default ProductItem;
