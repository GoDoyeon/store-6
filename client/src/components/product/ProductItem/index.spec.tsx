import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import ProductItem from './index';

describe('<ProductItem />', () => {
  it('should render same with snapshot', () => {
    const { container } = render(
      <ProductItem idx={1} thumbnail="test" title="test" price="2000" />,
    );
    expect(container).toMatchSnapshot();
  });
});
