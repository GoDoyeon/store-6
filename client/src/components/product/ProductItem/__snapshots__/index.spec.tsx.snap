// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> should render same with snapshot 1`] = `
.c1 {
  position: relative;
  height: 300px;
  width: 230px;
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.c3 {
  height: 300px;
  width: 230px;
  border-radius: 10px;
  -webkit-filter: drop-shadow(0px 4px 4px rgba(0,0,0,0.25)) drop-shadow(2px 2px 12px rgba(0,0,0,0.15));
  filter: drop-shadow(0px 4px 4px rgba(0,0,0,0.25)) drop-shadow(2px 2px 12px rgba(0,0,0,0.15));
  -webkit-transition: all 300ms;
  transition: all 300ms;
}

.c0:hover .c2 {
  -webkit-filter: brightness(60%);
  filter: brightness(60%);
  -webkit-transform: scale(1.08);
  -ms-transform: scale(1.08);
  transform: scale(1.08);
  box-shadow: 0px 4px 4px rgba(0,0,0,0.25), 2px 2px 12px rgba(0,0,0,0.15);
}

<div>
  <div
    class="c0 c1"
  >
    <img
      class="c2 c3"
      src="test"
    />
  </div>
</div>
`;
